Bluetooth: Host: Fix GATT write function for GAP Device Name characteristic

theob-pro opened the pull request:
The write function of the GAP Device Name characteristic was not adding the null character at the end of the received value.

See commits messages for details.

theob-pro committed:
Bluetooth: Host: Specify parameter needs of `bt_set_name`

State that the name given to `bt_set_name` must be null terminated.

If not, the behavior of the function may be wrong as it's using `strlen`
to get the length of the given name and `strcmp` to compare it to the
current device name. Both of those functions expect null terminated string.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Tests: Bluetooth: Add 'device_name' GATT test

The goal of this test is to ensure that setting the device name by
writing to the GAP 'Device Name' characteristic work as expected.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Host: GAP Device Name write now add null char at the end

The function used to write the value of the GAP Device Name
characteristic now ensure that the string passed to `bt_set_name` is
null terminated.

Also fix a wrong offset calculation.

The function used to write the value of the GAP Device Name
characteristic was returning an error when the offset + the length of
data to write was superior **or equal** to the maximum size of the
device name.

This caused the actual maximum device name size to be reduced by 1 byte.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Tests: Bluetooth: Add MTU exchange procedure in testlib

Add `bt_testlib_att_exchange_mtu` function to simplify the procedure.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

jhedberg commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -268,7 +268,7 @@ bool bt_is_ready(void);
  * @sa @kconfig{CONFIG_BT_DEVICE_NAME_MAX}.
  *
- * @param name New name
+ * @param name New name, must be null terminated
```
I'm not going to oppose this, but it feels a bit superfluous. We also don't have such a documentation convention elsewhere (AFAIK) like here: https://github.com/zephyrproject-rtos/zephyr/blob/4e201f21c8a7f58698e68a5f1cca72433f9778ac/include/zephyr/kernel.h#L1293

If an API takes a string parameter, and doesn't have any additional parameters that could be used to indicate the length, then the string parameter can be assumed to be a normal C string, i.e. null-terminated.

theob-pro replied:
I agree with you and that was my understanding of the API too. But the bug is a proof that it need to be stated.

The C strings are not trivial, we should always be careful when manipulating them. We may be used to them but not everyone is and they are definitely not the norm among the programming language. So I think being explicit in the documentation is important.

Thalley replied:
I would rather document this in a generic way: When `char *` is used we should always expect/assume that it is a NULL terminated string. This should not be an API or BT specific thing, but a Zephyr-wise assumption/expectation IMO

theob-pro replied:
The risk with Zephyr-wise assumption is that some people miss it... At least here in the documentation, if someone use the function it will be remained of the expectation.

Thalley replied:
> The risk with Zephyr-wise assumption is that some people miss it...

Wouldn't the same people miss it for the `strxxx` functions like `strlen` as well?

theob-pro replied:
Probably, yes. But is that a reason for not trying to make life easier on our side? 😅

@jhedberg and you seems to be against that so if you really want me to remove it, I will do it

Thalley commented on tests/bsim/bluetooth/host/gatt/device_name/src/common.h:
```diff
@@ -0,0 +1,24 @@
+#define ADVERTISER_NAME      "Advertiser Pro II"
+#define SERVER_ORIGINAL_NAME "Server Super Name"
+
+static char server_new_name[CONFIG_BT_DEVICE_NAME_MAX];
+
+static void generate_name(char *name, size_t length)
+{
+	for (int i = 0; i < length; i++) {
+		name[i] = (i % 26) + 97;
+	}
+}
+
```
Since you are defining `server_new_name` as a `char`, shouldn't you make room for and add the NULL terminator? That would make it possible to use `strlen(server_new_name)` instead, and treat it as a string here in the test

theob-pro replied:
No, my goal here is to check that even with a non-null terminated string the GATT write work fine. I should maybe have been more explicit 😅

Thalley replied:
> No, my goal here is to check that even with a non-null terminated string the GATT write work fine. I should maybe have been more explicit 😅

That should still work as expected when `server_new_name` is NULL terminated. When do you perform the write and use `strlen` then you'll only send the string as non-NULL terminated over air. 

My point was to avoid having a non-NULL terminated string value stored in a `char` array.

theob-pro replied:
I see what you mean, would it be better to have `server_new_name` be a byte array then?

Thalley replied:
> I see what you mean, would it be better to have server_new_name be a byte array then?

That is one option, yeah :) 

I don't see any issues storing it as a NULL-terminated string though. It doesn't affect the test, but allows the test to use string functions.

theob-pro replied:
I updated to use `uint8_t`. It's just a personal opinion but I feel the message of the test is better communicated without the NULL char

sjanc reviewed (commented):
value is already zero initialized and it will it will always have at least one last byte null (since check len >=) so string is null-terminated. Just in worst case it will reject peer writing max supported name len.

Also I think that 2 and 3 commits should be squashed, since otherwise there is off-by-one error introduced (might be accidentally backported/cherry-picked without patch that increased buffer)

theob-pro commented:
> value is already zero initialized

I am not certain that the `{}` syntax is supported by all compiler, @alwa-nordic do you know more on that?

> it will always have at least one last byte null (since check len >=) so string is null-terminated. Just in worst case it will reject peer writing max supported name len.

But this is wrong, the maximum size allowed by the spec is `CONFIG_BT_DEVICE_NAME_MAX`, we should not keep one byte from it for the null char.

> Also I think that 2 and 3 commits should be squashed, since otherwise there is off-by-one error introduced (might be accidentally backported/cherry-picked without patch that increased buffer)

I think you are right, I'll squash them!

sjanc commented:
> But this is wrong, `CONFIG_BT_DEVICE_NAME_MAX` doesn't include the null char in its value, we should not keep one byte from it for the null char.

Yes, as mentioned that max length name will be (incorrectly) rejected but there is no miscall to bt_set_name() as buffer is always NULL terminated

Thalley commented on tests/bsim/bluetooth/host/gatt/device_name/src/common.h:
```diff
@@ -0,0 +1,24 @@
+
+static void generate_name(char *name, size_t length)
+{
+	for (int i = 0; i < length; i++) {
+		name[i] = (i % 26) + 97;
+	}
```
```suggestion
	for (size_t i = 0; i < length; i++) {
		name[i] = (i % 26) + 97;
	}
```

How do we handle cases where `length` >= ARRAY_SIZE(server_new_name)`? 
Test case failure/assert?

theob-pro replied:
Will update the `size_t`.

> How do we handle cases where length >= ARRAY_SIZE(server_new_name)`?
Test case failure/assert?

Sorry I don't get it? It's not something that can happen in the test as it is right now, am I missing something?

Thalley replied:
> Sorry I don't get it? It's not something that can happen in the test as it is right now, am I missing something?

It's not an issue in this PR, but when someone modifies or adds additional tests in the future, then this function may be called with a length value that is larger than the array. 

I suggest to make it fail loudly instead of risking overwriting some other RAM in that case, which is much harder to debug

theob-pro replied:
Oh I think I understand the misunderstanding. The function is not updating `server_new_name` directly, instead it takes a buffer and the length of it

Thalley replied:
Oh... Yeah, my mind completely messed those 2 up :D Please ignore everything I've said :D

Thalley commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,32 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
 	mem_xor_n(dst, src1, src2, 16);
 }
 
+/**
+ * @brief Compare memory areas and their length
+ *
+ * If the length are 0, return true.
+ *
+ * @param m1 First memory area to compare, cannot be NULL even if length is 0
+ * @param len1 Length of the first memory area to compare
+ * @param m2 Second memory area to compare, cannot be NULL even if length is 0
+ * @param len2 Length of the second memory area to compare
+ *
+ * @returns true if both the length of the memory areas and their content are
+ * equal else false
+ */
+static inline bool memeq(const void *m1, size_t len1, const void *m2, size_t len2)
+{
+	int ret;
+
+	if (len1 != len2) {
+		return false;
+	}
+
+	ret = memcmp(m1, m2, len1);
+
+	return ret == 0;
+}
```
In either case, it should be added with unit tests. Let me know if you plan on keeping it in this PR or move to a new PR; I might have some additional review comments for it ;)

theob-pro replied:
I agree for the unit tests, I'll write some!

I was considering to make it its own PR but I have been told that we should not have API without user. Does you think unit tests would be good enough as "user" of the API?

Thalley replied:
> I was considering to make it its own PR but I have been told that we should not have API without user. Does you think unit tests would be good enough as "user" of the API?

There are probably a few hundred existing cases in Zephyr that could use this API :D Comparing memory is a pretty common operation

theob-pro replied:
What I mean is that in a PR from what I understood we should not introduce an API that is not directly used. But maybe I am wrong 😅

Thalley replied:
> What I mean is that in a PR from what I understood we should not introduce an API that is not directly used.

My point is that we have at least 120 existing cases in Zephyr that could be converted to use `memeq`, and any one of those, or all of those, could be converted by the addition here. That's similar to what I did in https://github.com/zephyrproject-rtos/zephyr/pull/84022 where I added the new API and then converted a few cases to add a user while keeping the PR fairly minimal (easier to review)

I am also OK with keeping it in this PR if you prefer.

Thalley commented on include/zephyr/sys/util.h:
```diff
@@ -783,6 +784,32 @@ static inline void mem_xor_128(uint8_t dst[16], const uint8_t src1[16], const ui
+{
+	int ret;
+
+	if (len1 != len2) {
+		return false;
+	}
```
I don't understand the purpose of the 2nd `len` field. If the caller knows that they 2 inputs are different lengths, why would they even call this function?

My suggestion would be to remove the 2nd `len` argument and just make it 
```
static inline bool memeq(const void *m1, const void *m2, size_t len)
{
    return memcmp(m1, m2, len) == 0;
}

```

I also believe that would make it easier for the compiler to optimize, but without having any data to support that claim.

theob-pro replied:
Comparing the length of the input before doing a `memcmp` is a common operation. The goal of that function is about not having to do the comparison yourself as it's common. I don't really see the gain of just doing a `memcmp` that return a boolean to be honnest

Thalley replied:
> I don't really see the gain of just doing a `memcmp` that return a boolean to be honnest

That's also the reason why the function doesn't exist in the standard library :D 

> Comparing the length of the input before doing a memcmp is a common operation. 

Sure, it's common, but not _that_ common. Looking over the uses of `memcmp.* == 0` seems to show that less than 10% of uses does a size comparison first. The most common usecase for this is comparing 2 statically sized values, and adding the len check, especially for an inline function, significantly increases memory use. 

If we move this to be a non-inline function, I'd be more OK with leaving the length check in, but it does still add a check that's mostly not necessary

theob-pro replied:
> That's also the reason why the function doesn't exist in the standard library :D

You have a point, maybe I could add both function, would you be fine with that?

> The most common usecase for this is comparing 2 statically sized values, and adding the len check, especially for an inline function, significantly increases memory use.

I understand your concern but I am pretty sure that the compiler will not really inline it (at least from my test with GCC I can tell you that the function is not inlined). Really I put it inline because this is how it's done in this file and without it the compiler complain that it's not used 😅

Thalley replied:
> I understand your concern but I am pretty sure that the compiler will not really inline it (at least from my test with GCC I can tell you that the function is not inlined). Really I put it inline because this is how it's done in this file and without it the compiler complain that it's not used

Might be true :D I've given up trying to follow how compilers do things these days :P 

> You have a point, maybe I could add both function, would you be fine with that?

I guess :) I will defer that to the maintainers of `sys`

pdgendt commented on tests/bsim/bluetooth/host/gatt/device_name/src/server.c:
```diff
@@ -0,0 +1,147 @@
+	LOG_DBG("Original Device Name: %s", original_name);
+	LOG_DBG("New Device Name: %s", new_name);
+
+	names_are_matching =
+		memeq(expected_name, sizeof(expected_name), new_name, strlen(new_name));
```
Creating https://github.com/zephyrproject-rtos/zephyr/pull/84159 only to provide for a single test here, is overkill.

theob-pro replied:
![image](https://github.com/user-attachments/assets/eab23953-b341-4a33-b5c6-896afdd5fdbb)

As pointed by @Thalley `memcmp.* == 0` is commonly used and according to him 10% of those usage check the size before. So I would say that there is a use case.

pdgendt replied:
BTW, `str(n)cmp` can be used here.

```suggestion
	names_are_matching = strncmp(expected_name, new_name, CONFIG_BT_DEVICE_NAME_MAX);
```

pdgendt replied:
So if it should be tested that is can't be greater than `CONFIG_BT_DEVICE_NAME_MAX` you already have a bigger problem with using `strlen`, as this can go out-of-bounds.

```c
	TEST_ASSERT(new_name[CONFIG_BT_DEVICE_NAME_MAX] == '\0', "new_name not NULL-terminated");
	names_are_matching = strncmp(expected_name, new_name, CONFIG_BT_DEVICE_NAME_MAX);
```

theob-pro replied:
That's more complicated. `memeq` is ensuring that the two lengths are equal.

Can we just have `memeq`, it's easier to use, makes the code shorter, makes it safer by avoiding reading in out-of-bound memory.

theob-pro commented:
Removed `memeq` in `include/zephyr/sys/util.h` and added it directly in the test file until https://github.com/zephyrproject-rtos/zephyr/pull/84159 move forward

Thalley commented on tests/bsim/bluetooth/host/gatt/device_name/src/server.c:
```diff
@@ -0,0 +1,161 @@
+		memeq(expected_name, sizeof(expected_name), new_name, strlen(new_name));
+	TEST_ASSERT(names_are_matching,
+		    "The name of the server doesn't match the one set by the client (server name: "
+		    "`%s`, expected name: `%s`)",
+		    new_name, expected_name);
```
`expected_name` may not be a NULL terminated string, and thus we should not treat it as such with `%s`

Thalley commented on tests/bsim/bluetooth/host/gatt/device_name/src/server.c:
```diff
@@ -0,0 +1,161 @@
+
+	bool names_are_matching;
+
+	uint8_t expected_name[CONFIG_BT_DEVICE_NAME_MAX];
```
What is the reasoning for why `expected_name` isn't stored a NULL terminated string like the others?

theob-pro replied:
So the goal by sending a byte array instead of a string is to ensure that we will correctly add the NULL byte, even if we send one without it. The GATT write procedure is taking a byte array, not a string so this may happen.
Here `expected_name` is a byte array, I agree it could be a string. But I prefer it to be exactly the same as the value we are sending.
